docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -0,0 +1,679 @@
+
+We could allow placing `FeatureGuardAttribute` on the property to indicate that it should act as a guard for a particular feature. The intention with any of these approaches is for the guard to prevent analyzer warnings:
+
+```csharp
+if (Feature.IsSupported) {
+    APIWhichRequiresDynamicCode(); // No warnings
+}
+
+[RequiresDynamicCode("Does something with dynamic codegen")]
+static void APIWhichRequiresDynamicCode() {
+    // ...
+}
+```
+
+The attribute instance needs to reference the feature somehow, whether as:
+
+- a reference to the feature attribute:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard(typeof(RequiresDynamicCodeAttribute))]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  This tells the analyzer enough that it can treat this as a guard without any extra information.
+
+  The analyzer wouldn't know about the relationship between this check and `RuntimeFeature.IsDynamicCodeSupported` or `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, but encoding this relationship isn't strictly necessary if we are only interested in representing feature _guards_ via attributes.
+
+  On its own this is not enough for ILLink/ILCompiler to do branch elimination, because there's no tie to the feature switch name.
+
+- a reference to the feature name string:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard("System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported")]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  The analyzer would need to hard-code the fact that this string corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name.
+
+- a reference to the existing feature check property:
+
+  ```csharp
+  class Feature {
+      [FeatureGuard(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeSupported))]
+      public static bool IsSupported => RuntimeFeature.IsDynamicCodeSupported;
+  }
+  ```
+
+  The analyzer would need to hard-code the fact that `RuntimeFeature.IsDynamicCodeSupported` corresponds to `RequiresDynamicCodeAttribute`, unless we model this relationship via attributes that represent feature switches.
+
+  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name, assuming it has existing knowledge of the fact that `RuntimeFeature.IsDynamicCodeSupported` is contrelled by the feature switch named `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, either from a substitution XML or from a separate attribute that encodes this.
```

Comment:
```suggestion
  This would be sufficient for ILLink/ILCompiler to treat `IsSupported` as a constant based on the feature switch name, assuming it has existing knowledge of the fact that `RuntimeFeature.IsDynamicCodeSupported` is controlled by the feature switch named `"System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported"`, either from a substitution XML or from a separate attribute that encodes this.
```